cassovary-collections/src/main/scala/com/twitter/cassovary/collections/CSeq.scala

```diff
@@ -0,0 +1,166 @@
+  override def length: Int = underlying.length
+
+  override def apply(idx: Int): T = underlying.apply(idx)
+}
+
+abstract class CSeqFactory[@specialized(Int, Long) T] {
+  def ofDim(n: Int): CSeq[T]
+
+  def fromCSeq(arr: CSeq[T], from: Int, until: Int): CSeq[T]
+
+  def fromArray(arr: Array[T]): CSeq[T]
+
+  def empty: CSeq[T]
+}
+
+/**
+  * Companion object that should be used, when creating `CSeq`s.
+  */
+object CSeq {
+  def ofDim[@specialized(Int, Long) T](n: Int)(implicit cSeqFactory: CSeqFactory[T]) = {
+    cSeqFactory.ofDim(n)
+  }
+
+  def apply[@specialized(Int, Long) T](arr: Array[T])(implicit cSeqFactory: CSeqFactory[T]) = {
+    cSeqFactory.fromArray(arr)
+  }
+
+  def apply[@specialized(Int, Long) T](arr: Array[T], from: Int, until: Int)
+                                      (implicit cSeqFactory: CSeqFactory[T]): CSeq[T] = {
+    cSeqFactory.fromCSeq(apply(arr), from, until)
+  }
+
+
+  def apply[@specialized(Int, Long) T](arr: CSeq[T], from: Int, until: Int)
+                                      (implicit cSeqFactory: CSeqFactory[T]): CSeq[T] = {
+    cSeqFactory.fromCSeq(arr, from, until)
+  }
+
+  def empty[@specialized(Int, Long) T](implicit cSeqFactory: CSeqFactory[T]) = {
+    cSeqFactory.empty
+  }
+
+  object Implicits {
+    private[this] val counter = new AtomicInteger(0)
+
+    // We use fundep materialization to generate classes for specific types
+    implicit def materializeCSeqFactory[T]: CSeqFactory[T] = macro impl[T]
+    def impl[T: c.WeakTypeTag](c: scala.reflect.macros.Context): c.Expr[CSeqFactory[T]] = {
+      import c.universe._
+
+      // In Scala 2.10 we need to manually ensure that type names are not duplicated
+      val cur = counter.getAndIncrement()
+
+      val sym = c.weakTypeOf[T]
+      val className = newTypeName("CSeq" + sym.toString + "$" + cur)
+      val arrayWrapperName = newTypeName("CSeq" + sym.toString + "ArrayWrapper$" + cur)
+      val factoryClassName = newTypeName("CFactory" + sym.toString + "$" + cur)
+
+      // The code below works in Scala 2.11 without numbering instances manually
+      //
+      // val className = internal.reificationSupport.freshTypeName("CSeq" + sym.toString + "$")
+      // val arrayWrapperName = internal.reificationSupport.freshTypeName("CSeq" + sym.toSt ring +
+      //        "ArrayWrapper$")
+      // val factoryClassName = internal.reificationSupport.freshTypeName("CFactory" + sym
+      // .toString + "$")
+      val clazz = q"""
+        class $className(underlying: CSeq[$sym], from: Int, until: Int)
+            extends com.twitter.cassovary.collections.CSeq[$sym] {
+          override def apply(idx: Int) = {
+            if (idx >= length) {
+              throw new IndexOutOfBoundsException(idx.toString)
+            } else {
+              underlying.apply(from + idx)
+            }
+          }
+
+          override def length: Int = until - from
+        }
+
+        class $arrayWrapperName(underlying: Array[$sym])
```

Comment:
do we need this extra class, or can we just make a function here to create a CSeq, i.e.,
def fromArray(underlying: Array[$sym]) = new CSeq[$sym] { ... }

and use fromArray below instead of $arrayWrapperName ?